kube-controllers/pkg/controllers/node/hostendpoints.go

```diff
@@ -17,300 +17,425 @@ package node
+// 2. Clean up any HostEndpoints that are lingering after a node was deleted
+// 3. Go through each node and sync HostEndpoints for that node
+func (c *autoHostEndpointController) syncHostEndpoints() {
+	logrus.Infof("Syncing all HostEndpoints")
+	if c.syncStatus != bapi.InSync {
+		return
+	}
+
+	if !c.config.AutoHostEndpointConfig.AutoCreate {
+		// Create host endpoints is disabled, we need to delete all hostEndpoints that might still be left over after being created by this controller
+		for _, hep := range c.hostEndpointTracker.getAllHostEndpoints() {
+			err := c.deleteHostEndpoint(hep.Name)
 			if err != nil {
-				logrus.WithError(err).Warnf("failed to delete hostendpoint %q", hep.Name)
-				return err
+				logrus.WithError(err).Error("failed to delete host endpoint")
 			}
 		}
+
+		// We can skip to the rest, all hostEndpoints are deleted and we don't want to generate any new ones
+		return
 	}
-	return nil
-}
 
-// createUpdateAutohostendpoints creates or updates all auto hostendpoints.
-func (c *autoHostEndpointController) createUpdateAutohostendpoints(ctx context.Context) error {
-	for _, node := range c.nodeCache {
-		err := c.syncAutoHostendpoint(ctx, node)
-		if err != nil {
-			logrus.WithError(err).Warnf("failed to sync hostendpoint for node %q", node.Name)
-			return err
+	for nodeName := range c.hostEndpointTracker.hostEndpointsByNode {
+		if _, ok := c.nodeCache[nodeName]; !ok {
+			c.deleteHostEndpointsForNode(nodeName)
 		}
 	}
-	return nil
+
+	for _, node := range c.nodeCache {
+		c.syncHostEndpointsForNode(node.Name)
+	}
 }
 
-// syncAllAutoHostendpoints ensures that the expected auto hostendpoints exist,
-func (c *autoHostEndpointController) syncAllAutoHostendpoints(ctx context.Context) error {
-	for n := 1; n <= 5; n++ {
-		logrus.Debugf("syncing all hostendpoints. attempt #%v", n)
-		autoHeps, err := c.listAutoHostendpoints(ctx)
-		if err != nil {
-			logrus.WithError(err).Warn("failed to list hostendpoints")
-			time.Sleep(retrySleepTime)
-			continue
-		}
+// syncHostEndpointsForNode() sync HostEndpoints for the particular node. It does the following
+// 1. If the node was deleted and is no longer in the node cache, delete all HostEndpoints we have associated with the node,
+// 2. Create/Sync/Delete the default HostEndpoint based on the createDefaultHostEndpoint option
+// 3. Iterate over the Templates and create HostEndpoints for each template that matches the Node by nodeSelector
+// 4. Check that there are no extra HostEndpoints we created before that no longer match the template, we delete those HostEndpoints
+func (c *autoHostEndpointController) syncHostEndpointsForNode(nodeName string) {
+	node := c.nodeCache[nodeName]
+	if node == nil {
+		// Node has been deleted clean up all HostEndpoints associated with this node
+		c.deleteHostEndpointsForNode(nodeName)
+		c.hostEndpointTracker.deleteHostEndpointForNode(nodeName)
```

Comment:
Would it make sense for `c.deleteHostEndpointsForNode` to call `c.hostEndpointTracker.deleteHostEndpointForNode` itself? That way you don't need to remember to call both of these (I think they always need to be called together?)